sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -188,17 +188,30 @@ pi_result check_error(CUresult result, const char *function, int line,
 /// contexts to be restored by SYCL.
 class ScopedContext {
 public:
-  ScopedContext(pi_context ctxt) {
+  ScopedContext(pi_context ctxt) : device(nullptr) {
     if (!ctxt) {
       throw PI_ERROR_INVALID_CONTEXT;
     }
 
     set_context(ctxt->get());
   }
 
-  ScopedContext(CUcontext ctxt) { set_context(ctxt); }
+  ScopedContext(CUcontext ctxt) : device(nullptr) { set_context(ctxt); }
 
-  ~ScopedContext() {}
+  // Creating a scoped context from a device will simply use the primary
+  // context, this should be used when there is no other appropriate context,
+  // such as for the device infos.
+  ScopedContext(pi_device device) : device(device) {
+    CUcontext ctxt;
+    cuDevicePrimaryCtxRetain(&ctxt, device->get());
+
+    set_context(ctxt);
```

Comment:
Would it make sense to check if the current active context is for the wanted device, and if so keep it?

Reply:
It would but I don't think we can because it's possible for the active context to be a "dangling" deleted context, so trying to query it for its device would likely crash, this is the same reason why we stopped trying to save and restore existing contexts in the first place.

Reply:
Ah, the old zombie context. Fair point.